src/display/text_layer.js

```diff
@@ -528,20 +534,48 @@ var renderTextLayer = (function renderTextLayerClosure() {
     },
 
     _render: function TextLayer_render(timeout) {
-      var textItems = this._textContent.items;
-      var textStyles = this._textContent.styles;
-      for (var i = 0, len = textItems.length; i < len; i++) {
-        appendText(this, textItems[i], textStyles);
+      let capability = createPromiseCapability();
+      let pump;
+      if (this._textContent === undefined) {
+        pump = () => {
+          reader.read().then(({ value, done, }) => {
+            if (done) {
+              capability.resolve();
+              return;
+            }
+            Util.appendToArray(this._textContent.items, value.items);
```

Comment:
`Util.extendObj(styleCache, value.styles);` look good for caching. processItems is whatever you see in the appendText-loop logic https://github.com/mukulmishra18/pdf.js/blob/e5d6050d45602c82c9c1f46768cae17feb7db6b2/src/display/text_layer.js#L566-L568.